### PR TITLE
kubeadm: delete pre-pull ds and add pull images check in upgrade apply and upgrade node 

### LINKED ISSUE #2123: investigate if we should stop using a DaemonSet for upgrade prepull
investigate if we should stop using a DaemonSet for upgrade prepull.

@rosti brought the idea that we might want to maintain a single peace of code for doing prepull during init/join/upgrade.

upgrade currently uses a separate method - a DS.

see the discussion here:
https://github.com/kubernetes/kubeadm/issues/1862#issuecomment-616899539

/kind feature
/kind design
/priority backlog

### PR DESCRIPTION
**Pull request by xlgao-zju**
Signed-off-by: Xianglin Gao <xianglin.gxl@alibaba-inc.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Delete pre-pull ds, Since we have already pulled images in the preflight:

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/preflight/checks.go#L835-L848

**Which issue(s) this PR fixes**:

Fixes [kubernetes/kubeadm#2123](https://github.com/kubernetes/kubeadm/issues/2123)

**Special notes for your reviewer**:
/assign @neolit123 
/assign @fabriziopandini 

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: do not use a DaemonSet for the pre-pull of control-plane images during "kubeadm upgrade apply". Individual node upgrades now pull the required images using a preflight check. The flag "--image-pull-timeout" for "kubeadm upgrade apply" is now deprecated and will be removed in a future release following a GA deprecation policy.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/upgrade/apply.go:91]: I think this should go through deprecation...

**Review by fabriziopandini — COMMENTED** [review]
@xlgao-zju thanks for this PR!
Might be I'm missing something, but IMO this PR is missing the alternative implementation of prepull during preflight checks (similar to what we are doing in init/join)

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (xlgao-zju) [reply cmd/kubeadm/app/cmd/upgrade/apply.go:91]: got it

**Comment by xlgao-zju** [comment]
> Might be I'm missing something, but IMO this PR is missing the alternative implementation of prepull during preflight checks (similar to what we are doing in init/join)

@fabriziopandini  You are right, I missed the alternative implementation(I thought we already have one), will push another commit soon.

**Comment by fabriziopandini** [comment]
/test pull-kubernetes-e2e-kind
@neolit123 is there a preflight check testing kubeadm upgrades?

**Comment by xlgao-zju** [comment]
@fabriziopandini
PR is updated, PTAL.

What did I update:
- mark `image-pull-timeout` flag  deprecated.
- add `RunPullImagesCheck()` for upgrade apply.
- add `kubeadm upgrade node phase preflight` , and in this phase , we will call `RunPullImagesCheck()` to pull images.

**Comment by xlgao-zju** [comment]
@fabriziopandini All test are passed, and I tested it locally, it worked fine. Please take another look.

**Comment by neolit123** [comment]
> @fabriziopandini All test are passed, and I tested it locally, it worked fine. Please take another look.

hi, did you test worker node / control-plane node upgrade as well?
we don't have a CI check for kubeadm upgrades.

**Review by neolit123 — COMMENTED** [review]
/cc @rosti

**Comment by xlgao-zju** [comment]
> hi, did you test worker node / control-plane node upgrade as well (kubeadm upgrade node)?
we don't have a CI check for kubeadm upgrades (only periodic checks).

run `kubeadm upgrade node` on control-plane node
```shell
# ./kubeadm upgrade node
[upgrade] Reading configuration from the cluster...
[upgrade] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[preflight] Running pre-flight checks
[preflight] Pulling images required for setting up a Kubernetes cluster
[preflight] This might take a minute or two, depending on the speed of your internet connection
[preflight] You can also perform this action in beforehand using 'kubeadm config images pull'
[upgrade] Upgrading your Static Pod-hosted control plane instance to version "v1.18.2"...
Static pod: kube-apiserver-izj6cgh7gke2gmwrnxbz3xz hash: 8f02a8f60b9df8e553e9c399767a4136
Static pod: kube-controller-manager-izj6cgh7gke2gmwrnxbz3xz hash: 77acc638a0fd5e0340bdc70827e58901
Static pod: kube-scheduler-izj6cgh7gke2gmwrnxbz3xz hash: 155707e0c19147c8dc5e997f089c0ad1
[upgrade/etcd] Upgrading to TLS for etcd
[upgrade/etcd] Non fatal issue encountered during upgrade: the desired etcd version "3.4.3-0" is not newer than the currently installed "3.4.3-0". Skipping etcd upgrade
[upgrade/staticpods] Writing new Static Pod manifests to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests070318394"
[upgrade/staticpods] Preparing for "kube-apiserver" upgrade
[upgrade/staticpods] Renewing apiserver certificate
[upgrade/staticpods] Renewing apiserver-kubelet-client certificate
[upgrade/staticpods] Renewing front-proxy-client certificate
[upgrade/staticpods] Renewing apiserver-etcd-client certificate
[upgrade/staticpods] Moved new manifest to "/etc/kubernetes/manifests/kube-apiserver.yaml" and backed up old manifest to "/etc/kubernetes/tmp/kubeadm-backup-manifests-2020-05-07-21-38-47/kube-apiserver.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[upgrade/staticpods] This might take a minute or longer depending on the component/version gap (timeout 5m0s)
Static pod: kube-apiserver-izj6cgh7gke2gmwrnxbz3xz hash: 8f02a8f60b9df8e553e9c399767a4136
Static pod: kube-apiserver-izj6cgh7gke2gmwrnxbz3xz hash: 8f02a8f60b9df8e553e9c399767a4136
Static pod: kube-apiserver-izj6cgh7gke2gmwrnxbz3xz hash: ec2c7d61dc35733a70889dd9a056e30e
[apiclient] Found 1 Pods for label selector component=kube-apiserver
[upgrade/staticpods] Component "kube-apiserver" upgraded successfully!
[upgrade/staticpods] Preparing for "kube-controller-manager" upgrade
[upgrade/staticpods] Renewing controller-manager.conf certificate
[upgrade/staticpods] Moved new manifest to "/etc/kubernetes/manifests/kube-controller-manager.yaml" and backed up old manifest to "/etc/kubernetes/tmp/kubeadm-backup-manifests-2020-05-07-21-38-47/kube-controller-manager.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[upgrade/staticpods] This might take a minute or longer depending on the component/version gap (timeout 5m0s)
Static pod: kube-controller-manager-izj6cgh7gke2gmwrnxbz3xz hash: 77acc638a0fd5e0340bdc70827e58901
Static pod: kube-controller-manager-izj6cgh7gke2gmwrnxbz3xz hash: 465270b301d1652b89ac23286a4dfca8
[apiclient] Found 1 Pods for label selector component=kube-controller-manager
[upgrade/staticpods] Component "kube-controller-manager" upgraded successfully!
[upgrade/staticpods] Preparing for "kube-scheduler" upgrade
[upgrade/staticpods] Renewing scheduler.conf certificate
[upgrade/staticpods] Moved new manifest to "/etc/kubernetes/manifests/kube-scheduler.yaml" and backed up old manifest to "/etc/kubernetes/tmp/kubeadm-backup-manifests-2020-05-07-21-38-47/kube-scheduler.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[upgrade/staticpods] This might take a minute or longer depending on the component/version gap (timeout 5m0s)
Static pod: kube-scheduler-izj6cgh7gke2gmwrnxbz3xz hash: 155707e0c19147c8dc5e997f089c0ad1
Static pod: kube-scheduler-izj6cgh7gke2gmwrnxbz3xz hash: 4c27e818fb13ff2e3013f3125dee30b5
[apiclient] Found 1 Pods for label selector component=kube-scheduler
[upgrade/staticpods] Component "kube-scheduler" upgraded successfully!
[upgrade] The control plane instance for this node was successfully updated!
[kubelet-start] Downloading configuration for the kubelet from the "kubelet-config-1.18" ConfigMap in the kube-system namespace
[kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[upgrade] The configuration for this node was successfully updated!
[upgrade] Now you should go ahead and upgrade the kubelet package using your package manager.
```

run `kubeadm upgrade node` on worker node
```shell
./kubeadm upgrade node
[upgrade] Reading configuration from the cluster...
[upgrade] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[preflight] Skipping phase. Not a control plane node.
[upgrade] Skipping phase. Not a control plane node.
[kubelet-start] Downloading configuration for the kubelet from the "kubelet-config-1.18" ConfigMap in the kube-system namespace
[kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
[upgrade] The configuration for this node was successfully updated!
[upgrade] Now you should go ahead and upgrade the kubelet package using your package manager.
```

@neolit123 I tested the cases you mentioned, and it all worked fine.

File: cmd/kubeadm/app/cmd/phases/upgrade/node/preflight.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/phases/upgrade/node/preflight.go:50]: The preflight phase might also contain other checks. One notable check, that's missing and is applicable to both worker and control plane nodes is the root check.
Hence, we should probably not make this as central as it is and we should probably make the comment above state that.

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/upgrade/apply.go:155]: No need to do that actually. It was already done by the above call to `enforceRequirements`. The result is in `cfg.NodeRegistration.IgnorePreflightErrors`. It's stored as a normal slice, so you'll have to convert it to a set again.

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (rosti) [reply cmd/kubeadm/app/cmd/upgrade/apply.go:91]: I am not sure if we should keep this flag operational until we can remove it.
@neolit123 @fabriziopandini WDYT?

**Comment by neolit123** [comment]
thank you for confirming @xlgao-zju 
one request about the release note:
> kubeadm: delete pre-pull ds and add pull images check in upgrade apply and upgrade node

please modify it to:
> kubeadm: do not use a DaemonSet for the pre-pull of control-plane images during "kubeadm upgrade apply". Individual node upgrades now pull the required images using a preflight check. The flag "--image-pull-timeout" for "kubeadm upgrade apply" is now deprecated and will be removed in a future release following a GA deprecation policy.

leaving approve to @rosti as this was his idea.
/assign @rosti

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/upgrade/apply.go:94]: please add a comment on top of this deprecation line in the lines of:
```
// TODO: The flag was deprecated in 1.19; remove the flag following a GA deprecation policy of 12 months or 2 releases (whichever is longer)
```
xref https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli

**Review by rosti — COMMENTED** [review]
Thanks @xlgao-zju !
I was looking forward to this one.

We also need a more descriptive release note. This change has the downside of making upgrades of HA clusters slower. That's because previously we pre-pulled the images in parallel on all control plane nodes, while now they are pulled sequentially.

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/upgrade/apply.go:91]: technically the flag should be operational until we remove it, which might be a bit difficult given we are removing the backend that it targets.

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (xlgao-zju) [reply cmd/kubeadm/app/cmd/upgrade/apply.go:95]: added

File: cmd/kubeadm/app/cmd/upgrade/apply.go
Comment (xlgao-zju) [reply cmd/kubeadm/app/cmd/upgrade/apply.go]: done

File: cmd/kubeadm/app/cmd/phases/upgrade/node/preflight.go
Comment (xlgao-zju) [reply cmd/kubeadm/app/cmd/phases/upgrade/node/preflight.go]: done. If this is a control-plane node, the pull the basic images. And run other check on both worker and control plane nodes.

**Comment by xlgao-zju** [comment]
@rosti @neolit123 PR is updated.

**Review by rosti — APPROVED** [review]
Thanks @xlgao-zju !
I think it looks good to me.
/approve
/hold

@fabriziopandini @neolit123 feel free to LGTM and remove the hold when you are happy with it.
@detiber I am interested in your opinion on this. This will make upgrades of HA clusters a bit slower, but no more than the time it takes to create the same cluster from scratch.

**Comment by xlgao-zju** [comment]
@neolit123 @fabriziopandini PTAL

**Comment by neolit123** [comment]
i will have another review pass on monday.

**Comment by neolit123** [comment]
/lgtm
/priority important-longterm

/hold cancel
let's see how this behaves in CI.

**Comment by xlgao-zju** [comment]
@neolit123 do not know why Tide retested my PR several times...

**Comment by neolit123** [comment]
it's slow and maybe buggy. just leave it to automatically retests itself if the PR is lgtm/approved (fejtabot will automatically type `/retest` when needed).

**Pull request merged**
